lib/linter/file-report.js

```diff
@@ -342,73 +445,200 @@ function validateSuggestions(suggest, messages) {
+	 */
+	addRuleMessage(ruleId, severity, ...args) {
 		const descriptor = normalizeMultiArgReportCall(...args);
-		const messages = metadata.messageIds;
-		const { sourceCode } = metadata;
+		const ruleDefinition = this.#ruleMapper(ruleId);
+		const messages = ruleDefinition?.meta?.messages;
 
 		assertValidNodeInfo(descriptor);
 
-		let computedMessage;
+		const computedMessage = computeMessageFromDescriptor(
+			descriptor,
+			messages,
+		);
 
-		if (descriptor.messageId) {
-			if (!messages) {
-				throw new TypeError(
-					"context.report() called with a messageId, but no messages were present in the rule metadata.",
-				);
-			}
-			const id = descriptor.messageId;
+		validateSuggestions(descriptor.suggest, messages);
 
-			if (descriptor.message) {
-				throw new TypeError(
-					"context.report() called with a message and a messageId. Please only pass one.",
-				);
-			}
-			if (!messages || !Object.hasOwn(messages, id)) {
-				throw new TypeError(
-					`context.report() called with a messageId of '${id}' which is not present in the 'messages' config: ${JSON.stringify(messages, null, 2)}`,
-				);
-			}
-			computedMessage = messages[id];
-		} else if (descriptor.message) {
-			computedMessage = descriptor.message;
-		} else {
-			throw new TypeError(
-				"Missing `message` property in report() call; add a message that describes the linting problem.",
-			);
-		}
+		this.messages.push(
+			createProblem({
+				ruleId,
+				severity,
+				node: descriptor.node,
+				message: interpolate(computedMessage, descriptor.data),
+				messageId: descriptor.messageId,
+				loc: descriptor.loc
+					? normalizeReportLoc(descriptor)
+					: this.#sourceCode.getLoc(descriptor.node),
+				fix: this.#disableFixes
+					? null
+					: normalizeFixes(descriptor, this.#sourceCode),
+				suggestions: this.#disableFixes
+					? []
+					: mapSuggestions(descriptor, this.#sourceCode, messages),
+				language: this.#language,
+			}),
+		);
 
-		validateSuggestions(descriptor.suggest, messages);
+		return this.messages.at(-1);
+	}
 
-		return createProblem({
-			ruleId: metadata.ruleId,
-			severity: metadata.severity,
-			node: descriptor.node,
-			message: interpolate(computedMessage, descriptor.data),
-			messageId: descriptor.messageId,
-			loc: descriptor.loc
-				? normalizeReportLoc(descriptor)
-				: sourceCode.getLoc(descriptor.node),
-			fix: metadata.disableFixes
-				? null
-				: normalizeFixes(descriptor, sourceCode),
-			suggestions: metadata.disableFixes
-				? []
-				: mapSuggestions(descriptor, sourceCode, messages),
-			language: metadata.language,
-		});
-	};
+	/**
+	 * Adds an error message to the report. Meant to be called outside of rules.
+	 * @param {LintProblem} descriptor The descriptor for the error message.
+	 * @returns {LintMessage} The created problem object.
+	 */
+	addError(descriptor) {
+		this.messages.push(createLintingProblem(descriptor, 2));
+		return this.messages.at(-1);
+	}
```

Comment:
What do you think about passing `this.#language` from these methods to `createLintingProblem()` instead of expecting the language to be passed through these methods as a property in `descriptor`? It's easy to forget to pass `language` property to these methods because those errors are difficult to spot (to be caught by testing, it would need a test with a custom language targeting the specific lint message). Also, there would be less code when calling these methods. And there should be no need to pass the language to these methods since they already know it.